{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ce2cfe32",
   "metadata": {},
   "source": [
    "# VCF Updating rs ID\n",
    "Problem: Compare data from VCF and TXT files, with the VCF file being the main file. If the columns 1 (chromosome), 2 (position), and Alternative sequence columns of both files match, insert the rs_ID data from the existing variant column of the TXT file as the third column of the VCF file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "750e0bc3",
   "metadata": {},
   "source": [
    "## 1. Explore Main VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1be06e80",
   "metadata": {},
   "outputs": [],
   "source": [
    "from fuc import pyvcf\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "75ae8b16",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Construct VcfFrame from a VCF file\n",
    "vcf = pyvcf.VcfFrame.from_file('SP01_vep.vcf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "8f4d0bf8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10</td>\n",
       "      <td>94426</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=T|intron_variant|MODIFIER|TUBB8|ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>95429</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=A|upstream_gene_variant|MODIFIER|TUBB8|ENS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>120756</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=T|regulatory_region_variant|MODIFIER|||Reg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>126237</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=A|downstream_gene_variant|MODIFIER|IL9RP2|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>126402</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=C|non_coding_transcript_exon_variant|MODIF...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM     POS ID REF ALT    QUAL FILTER  \\\n",
       "0    10   94426  .   C   T   . . .      .   \n",
       "1    10   95429  .   C  CA   . . .      .   \n",
       "2    10  120756  .   A   T   . . .      .   \n",
       "3    10  126237  .   G   A   . . .      .   \n",
       "4    10  126402  .   T   C   . . .      .   \n",
       "\n",
       "                                                INFO  \n",
       "0  CSQ=T|intron_variant|MODIFIER|TUBB8|ENSG000001...  \n",
       "1  CSQ=A|upstream_gene_variant|MODIFIER|TUBB8|ENS...  \n",
       "2  CSQ=T|regulatory_region_variant|MODIFIER|||Reg...  \n",
       "3  CSQ=A|downstream_gene_variant|MODIFIER|IL9RP2|...  \n",
       "4  CSQ=C|non_coding_transcript_exon_variant|MODIF...  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check cvf\n",
    "vcf.df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "30b217bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHROM     object\n",
      "POS        int64\n",
      "ID        object\n",
      "REF       object\n",
      "ALT       object\n",
      "QUAL      object\n",
      "FILTER    object\n",
      "INFO      object\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "print(vcf.df.dtypes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40eba0f4",
   "metadata": {},
   "source": [
    "## 2. Explore Information File (TXT) for Updating"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d094d250",
   "metadata": {},
   "outputs": [],
   "source": [
    "txt_df = pd.read_csv('SP01_vep.txt', sep='\\t', header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "482296f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>BIOTYPE</th>\n",
       "      <th>...</th>\n",
       "      <th>MOTIF_POS</th>\n",
       "      <th>HIGH_INF_POS</th>\n",
       "      <th>MOTIF_SCORE_CHANGE</th>\n",
       "      <th>TRANSCRIPTION_FACTORS</th>\n",
       "      <th>PHENOTYPES</th>\n",
       "      <th>NMD</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>CADD_RAW</th>\n",
       "      <th>Mastermind_MMID3</th>\n",
       "      <th>GO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>.</td>\n",
       "      <td>10:94426-94426</td>\n",
       "      <td>T</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>ENSG00000173876</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000309812.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Oocyte_maturation_defect_2+ClinVar+rs10904045</td>\n",
       "      <td>-</td>\n",
       "      <td>4.434</td>\n",
       "      <td>0.110377</td>\n",
       "      <td>-</td>\n",
       "      <td>GO:0003924:GTPase_activity,GO:0005200:structur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>.</td>\n",
       "      <td>10:94426-94426</td>\n",
       "      <td>T</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>ENSG00000173876</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000332708.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Oocyte_maturation_defect_2+ClinVar+rs10904045</td>\n",
       "      <td>-</td>\n",
       "      <td>4.434</td>\n",
       "      <td>0.110377</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>.</td>\n",
       "      <td>10:94426-94426</td>\n",
       "      <td>T</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>RP11-631M21.6</td>\n",
       "      <td>ENSG00000237297</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000416477.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Oocyte_maturation_defect_2+ClinVar+rs10904045</td>\n",
       "      <td>-</td>\n",
       "      <td>4.434</td>\n",
       "      <td>0.110377</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>.</td>\n",
       "      <td>10:94426-94426</td>\n",
       "      <td>T</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>ENSG00000173876</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000447903.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Oocyte_maturation_defect_2+ClinVar+rs10904045</td>\n",
       "      <td>-</td>\n",
       "      <td>4.434</td>\n",
       "      <td>0.110377</td>\n",
       "      <td>-</td>\n",
       "      <td>GO:0003924:GTPase_activity,GO:0005200:structur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>.</td>\n",
       "      <td>10:95429-95429</td>\n",
       "      <td>A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>TUBB8</td>\n",
       "      <td>ENSG00000173876</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000309812.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>2.885</td>\n",
       "      <td>0.016923</td>\n",
       "      <td>-</td>\n",
       "      <td>GO:0003924:GTPase_activity,GO:0005200:structur...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 75 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  #Uploaded_variation        Location Allele              Consequence  \\\n",
       "0                   .  10:94426-94426      T           intron_variant   \n",
       "1                   .  10:94426-94426      T           intron_variant   \n",
       "2                   .  10:94426-94426      T  downstream_gene_variant   \n",
       "3                   .  10:94426-94426      T           intron_variant   \n",
       "4                   .  10:95429-95429      A    upstream_gene_variant   \n",
       "\n",
       "     IMPACT         SYMBOL             Gene Feature_type            Feature  \\\n",
       "0  MODIFIER          TUBB8  ENSG00000173876   Transcript  ENST00000309812.4   \n",
       "1  MODIFIER          TUBB8  ENSG00000173876   Transcript  ENST00000332708.5   \n",
       "2  MODIFIER  RP11-631M21.6  ENSG00000237297   Transcript  ENST00000416477.1   \n",
       "3  MODIFIER          TUBB8  ENSG00000173876   Transcript  ENST00000447903.2   \n",
       "4  MODIFIER          TUBB8  ENSG00000173876   Transcript  ENST00000309812.4   \n",
       "\n",
       "                  BIOTYPE  ... MOTIF_POS HIGH_INF_POS MOTIF_SCORE_CHANGE  \\\n",
       "0          protein_coding  ...         -            -                  -   \n",
       "1          protein_coding  ...         -            -                  -   \n",
       "2  unprocessed_pseudogene  ...         -            -                  -   \n",
       "3          protein_coding  ...         -            -                  -   \n",
       "4          protein_coding  ...         -            -                  -   \n",
       "\n",
       "  TRANSCRIPTION_FACTORS                                     PHENOTYPES NMD  \\\n",
       "0                     -  Oocyte_maturation_defect_2+ClinVar+rs10904045   -   \n",
       "1                     -  Oocyte_maturation_defect_2+ClinVar+rs10904045   -   \n",
       "2                     -  Oocyte_maturation_defect_2+ClinVar+rs10904045   -   \n",
       "3                     -  Oocyte_maturation_defect_2+ClinVar+rs10904045   -   \n",
       "4                     -                                              -   -   \n",
       "\n",
       "  CADD_PHRED  CADD_RAW Mastermind_MMID3  \\\n",
       "0      4.434  0.110377                -   \n",
       "1      4.434  0.110377                -   \n",
       "2      4.434  0.110377                -   \n",
       "3      4.434  0.110377                -   \n",
       "4      2.885  0.016923                -   \n",
       "\n",
       "                                                  GO  \n",
       "0  GO:0003924:GTPase_activity,GO:0005200:structur...  \n",
       "1                                                  -  \n",
       "2                                                  -  \n",
       "3  GO:0003924:GTPase_activity,GO:0005200:structur...  \n",
       "4  GO:0003924:GTPase_activity,GO:0005200:structur...  \n",
       "\n",
       "[5 rows x 75 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "txt_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "acc5163b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['#Uploaded_variation', 'Location', 'Allele', 'Consequence', 'IMPACT',\n",
       "       'SYMBOL', 'Gene', 'Feature_type', 'Feature', 'BIOTYPE', 'EXON',\n",
       "       'INTRON', 'HGVSc', 'HGVSp', 'cDNA_position', 'CDS_position',\n",
       "       'Protein_position', 'Amino_acids', 'Codons', 'Existing_variation',\n",
       "       'REF_ALLELE', 'UPLOADED_ALLELE', 'DISTANCE', 'STRAND', 'FLAGS',\n",
       "       'SYMBOL_SOURCE', 'HGNC_ID', 'MANE_SELECT', 'MANE_PLUS_CLINICAL', 'TSL',\n",
       "       'APPRIS', 'CCDS', 'SIFT', 'PolyPhen', 'AF', 'AFR_AF', 'AMR_AF',\n",
       "       'EAS_AF', 'EUR_AF', 'SAS_AF', 'gnomADe_AF', 'gnomADe_AFR_AF',\n",
       "       'gnomADe_AMR_AF', 'gnomADe_ASJ_AF', 'gnomADe_EAS_AF', 'gnomADe_FIN_AF',\n",
       "       'gnomADe_NFE_AF', 'gnomADe_OTH_AF', 'gnomADe_SAS_AF', 'gnomADg_AF',\n",
       "       'gnomADg_AFR_AF', 'gnomADg_AMI_AF', 'gnomADg_AMR_AF', 'gnomADg_ASJ_AF',\n",
       "       'gnomADg_EAS_AF', 'gnomADg_FIN_AF', 'gnomADg_MID_AF', 'gnomADg_NFE_AF',\n",
       "       'gnomADg_OTH_AF', 'gnomADg_SAS_AF', 'CLIN_SIG', 'SOMATIC', 'PHENO',\n",
       "       'PUBMED', 'MOTIF_NAME', 'MOTIF_POS', 'HIGH_INF_POS',\n",
       "       'MOTIF_SCORE_CHANGE', 'TRANSCRIPTION_FACTORS', 'PHENOTYPES', 'NMD',\n",
       "       'CADD_PHRED', 'CADD_RAW', 'Mastermind_MMID3', 'GO'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#checks all column\n",
    "txt_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5b1540e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>REF_ALLELE</th>\n",
       "      <th>Existing_variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10:94426-94426</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs10904045,COSV59118495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10:94426-94426</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs10904045,COSV59118495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10:94426-94426</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs10904045,COSV59118495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10:94426-94426</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs10904045,COSV59118495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10:95429-95429</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>rs35442274</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Location Allele REF_ALLELE       Existing_variation\n",
       "0  10:94426-94426      T          C  rs10904045,COSV59118495\n",
       "1  10:94426-94426      T          C  rs10904045,COSV59118495\n",
       "2  10:94426-94426      T          C  rs10904045,COSV59118495\n",
       "3  10:94426-94426      T          C  rs10904045,COSV59118495\n",
       "4  10:95429-95429      A          -               rs35442274"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Colume Allele here is ALT\n",
    "txt_df = txt_df[['Location', 'Allele', 'REF_ALLELE', 'Existing_variation']]\n",
    "txt_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "460dd359",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Location              object\n",
      "Allele                object\n",
      "REF_ALLELE            object\n",
      "Existing_variation    object\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "print(txt_df.dtypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1746b694",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('10', '94426')"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#the location need to be separated\n",
    "from rsUpdate_vcf import split_location\n",
    "\n",
    "#test\n",
    "split_location('10:94426-94426')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "423a1848",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Split 'Location' txt using : into 2 new columns 'CHROM', 'POS'\n",
    "txt_df = txt_df.assign(\n",
    "    CHROM=txt_df['Location'].apply(lambda x: split_location(x)[0]),\n",
    "    POS=txt_df['Location'].apply(lambda x: split_location(x)[1]),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "215f7164",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>REF_ALLELE</th>\n",
       "      <th>Existing_variation</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10:94426-94426</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs10904045,COSV59118495</td>\n",
       "      <td>10</td>\n",
       "      <td>94426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10:94426-94426</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs10904045,COSV59118495</td>\n",
       "      <td>10</td>\n",
       "      <td>94426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10:94426-94426</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs10904045,COSV59118495</td>\n",
       "      <td>10</td>\n",
       "      <td>94426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10:94426-94426</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs10904045,COSV59118495</td>\n",
       "      <td>10</td>\n",
       "      <td>94426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10:95429-95429</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>rs35442274</td>\n",
       "      <td>10</td>\n",
       "      <td>95429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1084226</th>\n",
       "      <td>X:155238653-155238653</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>rs2472749</td>\n",
       "      <td>X</td>\n",
       "      <td>155238653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1084227</th>\n",
       "      <td>X:155238653-155238653</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>rs2472749</td>\n",
       "      <td>X</td>\n",
       "      <td>155238653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1084228</th>\n",
       "      <td>X:155238653-155238653</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>rs2472749</td>\n",
       "      <td>X</td>\n",
       "      <td>155238653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1084229</th>\n",
       "      <td>X:155238653-155238653</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>rs2472749</td>\n",
       "      <td>X</td>\n",
       "      <td>155238653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1084230</th>\n",
       "      <td>X:155238653-155238653</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>rs2472749</td>\n",
       "      <td>X</td>\n",
       "      <td>155238653</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1084231 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Location Allele REF_ALLELE       Existing_variation  \\\n",
       "0               10:94426-94426      T          C  rs10904045,COSV59118495   \n",
       "1               10:94426-94426      T          C  rs10904045,COSV59118495   \n",
       "2               10:94426-94426      T          C  rs10904045,COSV59118495   \n",
       "3               10:94426-94426      T          C  rs10904045,COSV59118495   \n",
       "4               10:95429-95429      A          -               rs35442274   \n",
       "...                        ...    ...        ...                      ...   \n",
       "1084226  X:155238653-155238653      T          A                rs2472749   \n",
       "1084227  X:155238653-155238653      T          A                rs2472749   \n",
       "1084228  X:155238653-155238653      T          A                rs2472749   \n",
       "1084229  X:155238653-155238653      T          A                rs2472749   \n",
       "1084230  X:155238653-155238653      T          A                rs2472749   \n",
       "\n",
       "        CHROM        POS  \n",
       "0          10      94426  \n",
       "1          10      94426  \n",
       "2          10      94426  \n",
       "3          10      94426  \n",
       "4          10      95429  \n",
       "...       ...        ...  \n",
       "1084226     X  155238653  \n",
       "1084227     X  155238653  \n",
       "1084228     X  155238653  \n",
       "1084229     X  155238653  \n",
       "1084230     X  155238653  \n",
       "\n",
       "[1084231 rows x 6 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "txt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ff903e1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'rs10904045'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#the rs ID need to be extracted from txt\n",
    "from rsUpdate_vcf import extract_id\n",
    "\n",
    "#test\n",
    "extract_id('rs10904045,COSV59118495')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f476d287",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          rs10904045\n",
       "1          rs10904045\n",
       "2          rs10904045\n",
       "3          rs10904045\n",
       "4          rs35442274\n",
       "              ...    \n",
       "1084226     rs2472749\n",
       "1084227     rs2472749\n",
       "1084228     rs2472749\n",
       "1084229     rs2472749\n",
       "1084230     rs2472749\n",
       "Name: ID, Length: 1084231, dtype: object"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "txt_df['ID'] = txt_df['Existing_variation'].apply(extract_id)\n",
    "txt_df['ID']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d646f7d",
   "metadata": {},
   "source": [
    "## 3. Updating main VCF file\n",
    "\n",
    "**Step 1:** Read and construct vcfFrame using pyvcf from the original VCF file.\n",
    "\n",
    "**Step 2:** Read the TXT file and extract only the desired columns: 'Location', 'Allele', 'REF_ALLELE', and 'Existing_variation'.\n",
    "\n",
    "Note:\n",
    "- Ensure 'Existing_variation' is of type str before using the extract_id function.\n",
    "- Ensure the 'POS' data in the TXT file is of type int, as it matches the type in the VCF file.\n",
    "\n",
    "**Step 3:** Merge the two DataFrames on the columns CHROM, POS, and ALT/Allele. Select relevant columns from the VCF DataFrame and fill empty cells with '.'.\n",
    "\n",
    "**Step 4:** Retrieve the header from the original VCF file.\n",
    "\n",
    "**Step 5:** Create the modified VCF file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2c8bf3e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/chiranan/Desktop/Python/GitHub/VCF_Test/rsUpdate_vcf.py:49: DtypeWarning: Columns (73,74) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  txt_df = pd.read_csv(f'{sp_num}_vep.txt', sep='\\t', header=0)\n",
      "/Users/chiranan/Desktop/Python/GitHub/VCF_Test/rsUpdate_vcf.py:49: DtypeWarning: Columns (69,70) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  txt_df = pd.read_csv(f'{sp_num}_vep.txt', sep='\\t', header=0)\n"
     ]
    }
   ],
   "source": [
    "from rsUpdate_vcf import main\n",
    "\n",
    "files_to_update = ['SP01', 'SP04', 'SP09', 'SP13', 'SP24', 'SP31']\n",
    "for file in files_to_update:\n",
    "    main(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "afec46df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10</td>\n",
       "      <td>94545</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=T|intron_variant|MODIFIER|TUBB8|ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>95032</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=T|intron_variant|MODIFIER|TUBB8|ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>95226</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=T|5_prime_UTR_variant|MODIFIER|TUBB8|ENSG0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>95229</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=T|5_prime_UTR_variant|MODIFIER|TUBB8|ENSG0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>95429</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=A|upstream_gene_variant|MODIFIER|TUBB8|ENS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434927</th>\n",
       "      <td>7</td>\n",
       "      <td>77958262</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=A|intron_variant|MODIFIER|MAGI2|ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434928</th>\n",
       "      <td>7</td>\n",
       "      <td>77958559</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=C|intron_variant|MODIFIER|MAGI2|ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434929</th>\n",
       "      <td>7</td>\n",
       "      <td>77961831</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=C|intron_variant|MODIFIER|MAGI2|ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434930</th>\n",
       "      <td>7</td>\n",
       "      <td>77967743</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=T|intron_variant|MODIFIER|MAGI2|ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434931</th>\n",
       "      <td>7</td>\n",
       "      <td>77968301</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AAAACAAACAAAC</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=AAACAAACAAAC|intron_variant|MODIFIER|MAGI2...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>434932 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHROM       POS ID REF            ALT    QUAL FILTER  \\\n",
       "0         10     94545  .   C              T   . . .      .   \n",
       "1         10     95032  .   C              T   . . .      .   \n",
       "2         10     95226  .   A              T   . . .      .   \n",
       "3         10     95229  .   C              T   . . .      .   \n",
       "4         10     95429  .   C             CA   . . .      .   \n",
       "...      ...       ... ..  ..            ...     ...    ...   \n",
       "434927     7  77958262  .   T              A   . . .      .   \n",
       "434928     7  77958559  .   T              C   . . .      .   \n",
       "434929     7  77961831  .   T              C   . . .      .   \n",
       "434930     7  77967743  .   G              T   . . .      .   \n",
       "434931     7  77968301  .   A  AAAACAAACAAAC   . . .      .   \n",
       "\n",
       "                                                     INFO  \n",
       "0       CSQ=T|intron_variant|MODIFIER|TUBB8|ENSG000001...  \n",
       "1       CSQ=T|intron_variant|MODIFIER|TUBB8|ENSG000001...  \n",
       "2       CSQ=T|5_prime_UTR_variant|MODIFIER|TUBB8|ENSG0...  \n",
       "3       CSQ=T|5_prime_UTR_variant|MODIFIER|TUBB8|ENSG0...  \n",
       "4       CSQ=A|upstream_gene_variant|MODIFIER|TUBB8|ENS...  \n",
       "...                                                   ...  \n",
       "434927  CSQ=A|intron_variant|MODIFIER|MAGI2|ENSG000001...  \n",
       "434928  CSQ=C|intron_variant|MODIFIER|MAGI2|ENSG000001...  \n",
       "434929  CSQ=C|intron_variant|MODIFIER|MAGI2|ENSG000001...  \n",
       "434930  CSQ=T|intron_variant|MODIFIER|MAGI2|ENSG000001...  \n",
       "434931  CSQ=AAACAAACAAAC|intron_variant|MODIFIER|MAGI2...  \n",
       "\n",
       "[434932 rows x 8 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#example of original SP13\n",
    "vcf_SP13_original = pyvcf.VcfFrame.from_file('SP13_vep.vcf')\n",
    "vcf_SP13_original.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "28c2b937",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10</td>\n",
       "      <td>94545</td>\n",
       "      <td>rs10904047</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=T|intron_variant|MODIFIER|TUBB8|ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>95032</td>\n",
       "      <td>rs181491095</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=T|intron_variant|MODIFIER|TUBB8|ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>95226</td>\n",
       "      <td>rs6560829</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=T|5_prime_UTR_variant|MODIFIER|TUBB8|ENSG0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>95229</td>\n",
       "      <td>rs6560830</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=T|5_prime_UTR_variant|MODIFIER|TUBB8|ENSG0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>95429</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=A|upstream_gene_variant|MODIFIER|TUBB8|ENS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434927</th>\n",
       "      <td>7</td>\n",
       "      <td>77958262</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=A|intron_variant|MODIFIER|MAGI2|ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434928</th>\n",
       "      <td>7</td>\n",
       "      <td>77958559</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=C|intron_variant|MODIFIER|MAGI2|ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434929</th>\n",
       "      <td>7</td>\n",
       "      <td>77961831</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=C|intron_variant|MODIFIER|MAGI2|ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434930</th>\n",
       "      <td>7</td>\n",
       "      <td>77967743</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=T|intron_variant|MODIFIER|MAGI2|ENSG000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434931</th>\n",
       "      <td>7</td>\n",
       "      <td>77968301</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AAAACAAACAAAC</td>\n",
       "      <td>. . .</td>\n",
       "      <td>.</td>\n",
       "      <td>CSQ=AAACAAACAAAC|intron_variant|MODIFIER|MAGI2...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>434932 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHROM       POS           ID REF            ALT    QUAL FILTER  \\\n",
       "0         10     94545   rs10904047   C              T   . . .      .   \n",
       "1         10     95032  rs181491095   C              T   . . .      .   \n",
       "2         10     95226    rs6560829   A              T   . . .      .   \n",
       "3         10     95229    rs6560830   C              T   . . .      .   \n",
       "4         10     95429            .   C             CA   . . .      .   \n",
       "...      ...       ...          ...  ..            ...     ...    ...   \n",
       "434927     7  77958262            .   T              A   . . .      .   \n",
       "434928     7  77958559            .   T              C   . . .      .   \n",
       "434929     7  77961831            .   T              C   . . .      .   \n",
       "434930     7  77967743            .   G              T   . . .      .   \n",
       "434931     7  77968301            .   A  AAAACAAACAAAC   . . .      .   \n",
       "\n",
       "                                                     INFO  \n",
       "0       CSQ=T|intron_variant|MODIFIER|TUBB8|ENSG000001...  \n",
       "1       CSQ=T|intron_variant|MODIFIER|TUBB8|ENSG000001...  \n",
       "2       CSQ=T|5_prime_UTR_variant|MODIFIER|TUBB8|ENSG0...  \n",
       "3       CSQ=T|5_prime_UTR_variant|MODIFIER|TUBB8|ENSG0...  \n",
       "4       CSQ=A|upstream_gene_variant|MODIFIER|TUBB8|ENS...  \n",
       "...                                                   ...  \n",
       "434927  CSQ=A|intron_variant|MODIFIER|MAGI2|ENSG000001...  \n",
       "434928  CSQ=C|intron_variant|MODIFIER|MAGI2|ENSG000001...  \n",
       "434929  CSQ=C|intron_variant|MODIFIER|MAGI2|ENSG000001...  \n",
       "434930  CSQ=T|intron_variant|MODIFIER|MAGI2|ENSG000001...  \n",
       "434931  CSQ=AAACAAACAAAC|intron_variant|MODIFIER|MAGI2...  \n",
       "\n",
       "[434932 rows x 8 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#example of updated SP13\n",
    "vcf_SP13_update = pyvcf.VcfFrame.from_file('SP13_updated.vcf')\n",
    "vcf_SP13_update.df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
